src/touchscreen.c

```diff
@@ -0,0 +1,157 @@
+static BOOL TouchscreenHitbox_PointIsInRect(const TouchscreenHitbox *hitbox, u32 x, u32 y);
+
+static int TouchscreenHitbox_FindRectAtPoint(const TouchscreenHitbox *hitboxes, u32 x, u32 y) {
+    // In this function and TouchscreenHitbox_PointIsInRect, GameFreak uses the bitwise-and operator (&) instead of the binary-and operator (&&).
+    // Other bad code decisions effectively mask this particular choice, albeit with a performance impact.
+    for (int i = 0; hitboxes[i].rect.top != TOUCHSCREEN_RECTLIST_END; ++i) {
+        if ((x - hitboxes[i].rect.left < hitboxes[i].rect.right - hitboxes[i].rect.left) & (y - hitboxes[i].rect.top < hitboxes[i].rect.bottom - hitboxes[i].rect.top)) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+static BOOL TouchscreenHitbox_PointIsInCircle(const TouchscreenHitbox *hitbox, u32 x, u32 y) {
+    x = (hitbox->circle.x - x) * (hitbox->circle.x - x);
+    y = (hitbox->circle.y - y) * (hitbox->circle.y - y);
+    return x + y < (hitbox->circle.r * hitbox->circle.r);
+}
+
+static BOOL TouchscreenHitbox_PointIsInRect(const TouchscreenHitbox *hitbox, u32 x, u32 y) {
+    return ((x - hitbox->rect.left < hitbox->rect.right - hitbox->rect.left) & (y - hitbox->rect.top < hitbox->rect.bottom - hitbox->rect.top)) ? TRUE : FALSE;
+}
+
+int TouchscreenHitbox_FindRectAtTouchHeld(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchHeld) {
+        return TouchscreenHitbox_FindRectAtPoint(hitboxes, gSystem.touchX, gSystem.touchY);
+    }
+
+    return -1;
+}
+
+int TouchscreenHitbox_FindRectAtTouchNew(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchNew) {
+        return TouchscreenHitbox_FindRectAtPoint(hitboxes, gSystem.touchX, gSystem.touchY);
+    }
+
+    return -1;
+}
+
+int TouchscreenHitbox_FindHitboxAtTouchHeld(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchHeld) {
+        for (int i = 0; hitboxes[i].circle.sentinel != TOUCHSCREEN_RECTLIST_END; ++i) {
+            if (hitboxes[i].circle.sentinel == TOUCHSCREEN_CIRCLE_SENTINEL) {
+                if (TouchscreenHitbox_PointIsInCircle(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
+                    return i;
+                }
+            } else {
+                if (TouchscreenHitbox_PointIsInRect(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
+                    return i;
+                }
+            }
+        }
+    }
+
+    return -1;
+}
+
+int TouchscreenHitbox_FindHitboxAtTouchNew(const TouchscreenHitbox *hitboxes) {
+    if (gSystem.touchNew) {
+        for (int i = 0; hitboxes[i].circle.sentinel != TOUCHSCREEN_RECTLIST_END; ++i) {
+            if (hitboxes[i].circle.sentinel == TOUCHSCREEN_CIRCLE_SENTINEL) {
+                if (TouchscreenHitbox_PointIsInCircle(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
+                    return i;
+                }
+            } else {
+                if (TouchscreenHitbox_PointIsInRect(&hitboxes[i], gSystem.touchX, gSystem.touchY)) {
```

Comment:
else if can be merged into one statement

Reply:
etc for where it looks like you copied retsam code for the rest of the file

Reply:
I think this layout communicates the intention more clearly.

Reply:
gotta agree here, should be else if